### PR TITLE
 Add abstraction for hosting Saturn on Azure Functions 

### PR DESCRIPTION
**Pull request by Krzysztof-Cieslak**

### CODE REVIEW

**Comment by johlrich** [comment]
I've done some Giraffe/Saturn expiramentation with Azure Functions as well, here's some feedback based on that:

I don't think the library should take a direct dependency on the FunctionsSdk. The sdk really just helps to help generate the function.json runtime metadata files based on your attributes, so it should probably only be a dependency of sample applications.

Since an Azure Function binding can request either type, I'd recommend using `ILogger` over `TraceWriter` to stick with the newer abstractions and to have structured logging support. 

Because the json/xml/content negotation bits introduced by the `AddGiraffe` call will be missing at runtime, you'll probably see errors when using many of the helpers (unless this was handled elsewhere that I missed). What I ended up doing here was create my own static instances and then use a thin wrapper over the `IServiceProvider` in the `HttpContext` to shim these services in. [Here's a gist of the adapter I'm using today](https://gist.github.com/johlrich/04bbaffb5a58cbce5e2535197053062d).

Merging this idea with your existing PR I could imagine this being supported with additional operations like `use_json_serializer`, `use_xml_serializer`, `use_negotation_config` on the CE and if these aren't set to just use the [default implementations](https://github.com/giraffe-fsharp/Giraffe/blob/master/src/Giraffe/Middleware.fs#L113-L116)

**Comment by Krzysztof-Cieslak** [comment]
Great, thanks a lot for all the feedback! :-)

**Comment by Krzysztof-Cieslak** [comment]
Applied all @johlrich suggestions. Review welcomed.

File: src/Saturn.AzureFunctions/AzureFunctions.fs [hunk 1]
```diff
@@ -5,39 +5,55 @@ open Giraffe
...
       let next : HttpContext -> Task<HttpContext option> = Some >> Task.FromResult
       fun req ->
+        req.HttpContext.RequestServices <- wrapGiraffeServices req.HttpContext.RequestServices
         let r =
           match state.Router with
           | Some r -> r
           | None -> failwith "Router needs to be defined"
         let r = router {
-          pipe_through (fun nxt ctx -> state.Logger |> Option.iter (fun log -> ctx.Items.["TraceWriter"] <- log); nxt ctx)
+          pipe_through (fun nxt ctx -> state.Logger |> Option.iter (fun log -> ctx.Items.["Logger"] <- log); nxt ctx)
```
Comment (johlrich) [inline src/Saturn.AzureFunctions/AzureFunctions.fs]: Hmm, I wonder if this module should also provide a helper or an extension on HttpContext to get the Logger being set here vs relying on people knowing and likely making their own helper. What do you think?

**Pull request merged**